contentcuration/contentcuration/static/js/exercise_creation/views.js

```diff
@@ -0,0 +1,562 @@
+
+    template: require("./hbtemplates/exercise_list.handlebars"),
+
+    render: function (){
+        this.$el.html(this.template({exercise_list: this.collection.toJSON()}))
+    },
+
+    add_exercise: function() {
+        this.collection.create();
+    },
+
+    delete_exercise: function(ev) {
+        var id = ev.currentTarget.value;
+        var model = this.collection.get(id);
+        model.destroy();
+        return false;
+    }
+});
+
+var FileUploadView = Backbone.View.extend({
+
+    initialize: function(options) {
+        _.bindAll(this, "file_uploaded");
+        this.callback = options.callback;
+        this.modal = options.modal;
+        this.render();
+    },
+
+    template: require("./hbtemplates/file_upload.handlebars"),
+
+    modal_template: require("./hbtemplates/file_upload_modal.handlebars"),
+
+    render: function() {
+
+        if (this.modal) {
+            this.$el.html(this.modal_template());
+            this.$(".modal-body").append(this.template());
+            $("body").append(this.el);
+            this.$(".modal").modal({show: true});
+            this.$(".modal").on("hide.bs.modal", this.close);
+        } else {
+            this.$el.html(this.template());
+        }
+
+        // TODO parameterize to allow different file uploads depending on initialization.
+        this.dropzone = new Dropzone(this.$("#dropzone").get(0), {maxFiles: 1, clickable: ["#dropzone", ".fileinput-button"], acceptedFiles: "image/*", url: window.Urls.file_upload(), headers: {"X-CSRFToken": get_cookie("csrftoken")}});
+        this.dropzone.on("success", this.file_uploaded);
+
+    },
+
+    file_uploaded: function(file) {
+        this.callback(JSON.parse(file.xhr.response).filename);
+        this.close();
+    },
+
+    close: function() {
+        if (this.modal) {
+            this.$(".modal").modal('hide');
+        }
+        this.remove();
+    }
+});
+
+var EditorView = Backbone.View.extend({
+
+    tagName: "div",
+
+    initialize: function(options) {
+        _.bindAll(this, "return_markdown", "add_image", "deactivate_editor", "activate_editor", "save_and_close", "save", "render");
+        this.edit_key = options.edit_key;
+        this.editing = false;
+        this.render();
+        this.listenTo(this.model, "change:" + this.edit_key, this.render);
+    },
+
+    events: {
+        "click .ql-image": "add_image_popup"
+    },
+
+    add_image_popup: function() {
+        var view = new FileUploadView({callback: this.add_image, modal: true});
+    },
+
+    add_image: function(filename) {
+        this.editor.insertEmbed(this.editor.getSelection() !== null ? this.editor.getSelection().start : this.editor.getLength(), "image", "/media/" + filename);
+        this.save();
+    },
+
+    edit_template: require("./hbtemplates/editor.handlebars"),
+
+    view_template: require("./hbtemplates/editor_view.handlebars"),
+
+    render: function() {
+        if (this.editing) {
+            if (!this.setting_model) {
```

Comment:
The view rerenders on model change. But, the save method below modifies the exact attribute that it is listening to. If we don't stop the rerender, we needlessly reparse the markdown to HTML into the editor. This led to some weird behaviour (due to race conditions) during manual testing, so I stopped it.

The only other alternative would be to do the ```set``` in the ```save``` method with ```{silent: true}``` as an option, but other behaviour relies on listening to the model's change events.